imessage/bluebubbles/api.go

```diff
@@ -0,0 +1,1631 @@
+package bluebubbles
+
+import (
+	"bytes"
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"math/rand"
+	"mime/multipart"
+	"net/http"
+	"net/url"
+	"os"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+	"unicode"
+
+	"github.com/gorilla/websocket"
+	"github.com/rs/zerolog"
+	"github.com/sahilm/fuzzy"
+	"maunium.net/go/mautrix/id"
+
+	"go.mau.fi/mautrix-imessage/imessage"
+)
+
+const (
+	// Ref: https://github.com/BlueBubblesApp/bluebubbles-server/blob/master/packages/server/src/server/events.ts
+	NewMessage            string = "new-message"
+	MessageSendError      string = "message-send-error"
+	MessageUpdated        string = "updated-message"
+	ParticipantRemoved    string = "participant-removed"
+	ParticipantAdded      string = "participant-added"
+	ParticipantLeft       string = "participant-left"
+	GroupIconChanged      string = "group-icon-changed"
+	GroupIconRemoved      string = "group-icon-removed"
+	ChatReadStatusChanged string = "chat-read-status-changed"
+	TypingIndicator       string = "typing-indicator"
+	GroupNameChanged      string = "group-name-change"
+	IMessageAliasRemoved  string = "imessage-alias-removed"
+)
+
+type blueBubbles struct {
+	bridge            imessage.Bridge
+	log               zerolog.Logger
+	ws                *websocket.Conn
+	messageChan       chan *imessage.Message
+	receiptChan       chan *imessage.ReadReceipt
+	typingChan        chan *imessage.TypingNotification
+	chatChan          chan *imessage.ChatInfo
+	contactChan       chan *imessage.Contact
+	messageStatusChan chan *imessage.SendMessageStatus
+	backfillTaskChan  chan *imessage.BackfillTask
+
+	contactsLastRefresh time.Time
+	contacts            []Contact
+
+	bbRequestLock sync.Mutex
+
+	usingPrivateApi bool
+}
+
+func NewBlueBubblesConnector(bridge imessage.Bridge) (imessage.API, error) {
+	return &blueBubbles{
+		bridge: bridge,
+		log:    bridge.GetZLog().With().Str("component", "bluebubbles").Logger(),
+
+		messageChan:       make(chan *imessage.Message, 256),
+		receiptChan:       make(chan *imessage.ReadReceipt, 32),
+		typingChan:        make(chan *imessage.TypingNotification, 32),
+		chatChan:          make(chan *imessage.ChatInfo, 32),
+		contactChan:       make(chan *imessage.Contact, 2048),
+		messageStatusChan: make(chan *imessage.SendMessageStatus, 32),
+		backfillTaskChan:  make(chan *imessage.BackfillTask, 32),
+	}, nil
+}
+
+func init() {
+	imessage.Implementations["bluebubbles"] = NewBlueBubblesConnector
+}
+
+func (bb *blueBubbles) Start(readyCallback func()) error {
+	bb.log.Trace().Msg("Start")
+
+	ws, _, err := websocket.DefaultDialer.Dial(bb.wsUrl(), nil)
```

Comment:
This should be moved to a separate function so it can reconnect automatically if the connection is temporarily lost (e.g. if the bluebubbles server is restarted)